kafka-ui-api/src/main/java/com/provectus/kafka/ui/serde/schemaregistry/BufAndSchemaRegistryAwareRecordSerDe.java

```diff
@@ -20,104 +35,319 @@
-      return this.schemaRegistryAwareRecordSerDe.deserialize(msg);
+    String keyType = null;
+    String valueType = null;
+
+    ProtoSchema protoSchemaFromTopic = protoSchemaFromTopic(msg.topic());
+
+    if (protoSchemaFromTopic != null) {
+      keyType = protoSchemaFromTopic.getFullyQualifiedTypeName();
+      valueType = protoSchemaFromTopic.getFullyQualifiedTypeName();
+    }
+
+    ProtoSchema protoSchemaForKeyFromHeader = protoValueSchemaFromHeaders(msg.headers());
+    if (protoSchemaForKeyFromHeader != null) {
+      keyType = protoSchemaForKeyFromHeader.getFullyQualifiedTypeName();
     }
 
-    protoSchema = protoSchemaFromTopic(msg.topic());
+    ProtoSchema protoSchemaFromHeader = protoKeySchemaFromHeaders(msg.headers());
+    if (protoSchemaFromHeader != null) {
+      valueType = protoSchemaFromHeader.getFullyQualifiedTypeName();
+    }
+
+    String keyTypeFromConfig = protobufKeyMessageNameByTopic.get(msg.topic());
+    if (keyTypeFromConfig != null) {
+      keyType = keyTypeFromConfig;
+    }
+
+    String valueTypeFromConfig = protobufMessageNameByTopic.get(msg.topic());
+    if (valueTypeFromConfig != null) {
+      valueType = valueTypeFromConfig;
+    }
 
-    if (protoSchema != null) {
-      // TODO: parse message and return it
-      log.info("Skipping buf for schema: {}", protoSchema.getFullyQualifiedTypeName());
-      return this.schemaRegistryAwareRecordSerDe.deserialize(msg);
+    if (valueType != null) {
+      return deserializeProto(msg, keyType, valueType);
     }
 
+    log.debug("No proto schema found, skipping buf for topic {}", msg.topic());
+
     return this.schemaRegistryAwareRecordSerDe.deserialize(msg);
   }
 
-  @Nullable ProtoSchema protoSchemaFromHeaders(Headers headers) {
-    // extract PROTOBUF_TYPE header
+  private DeserializedKeyValue deserializeProto(ConsumerRecord<Bytes, Bytes> msg,
+        String keyFullyQualifiedType,
+        String valueFullyQualifiedType) {
+    Descriptor keyDescriptor = null;
+    if (keyFullyQualifiedType != null) {
+      keyDescriptor = getDescriptor(keyFullyQualifiedType);
+      if (keyDescriptor == null) {
+        log.warn("No key descriptor found for topic {} with schema {}", msg.topic(), keyFullyQualifiedType);
+      }
+    }
+
+    Descriptor valueDescriptor = null;
+    if (valueFullyQualifiedType != null) {
+      valueDescriptor = getDescriptor(valueFullyQualifiedType);
+      if (valueDescriptor == null) {
+        log.warn("No value descriptor found for topic {} with schema {}", msg.topic(), valueFullyQualifiedType);
+      }
+    }
+
+    return this.deserializeProtobuf(msg, keyDescriptor, valueDescriptor);
+  }
+
+  @Nullable
+  ProtoSchema protoKeySchemaFromHeaders(Headers headers) {
+    // Get PROTOBUF_TYPE_KEY header.
     String fullyQualifiedTypeName = null;
-    for (Header header : headers.headers("PROTOBUF_TYPE")) {
+    for (Header header : headers.headers("PROTOBUF_TYPE_KEY")) {
```

Comment:
PROTOBUF_TYPE_KEY? This wasn't updated in the docs https://www.notion.so/clearstreet/Protobuf-on-Kafka-ac912401cf504c75956d907b86a12f55

Reply:
Updated the doc